[lld][WebAssembly] Avoid emitting empty __wasm_apply_data_relocs function

yamt opened the pull request:
Instead of always generating __wasm_apply_data_relocs when relevant options like -pie and -shared are specified, generate it only when the relevant relocations are actually necessary.

Note: omitting empty __wasm_apply_data_relocs is not a problem because the export is optional in the spec (DynamicLinking.md) and all runtime linker implementations I'm aware of implement it that way.  (emscripten, toywasm, wasm-tools)

Motivations:

* This possibly reduces the module size

* This is also a preparation to fix https://github.com/llvm/llvm-project/issues/107387, for which it isn't obvious if we need these relocations at the time of createSyntheticSymbols. (unless we introduce a new explicit option like --non-pie-dynamic-link.)

yamt committed:
[lld][WebAssembly]: Avoid emitting empty __wasm_apply_data_relocs function

Instead of always generating __wasm_apply_data_relocs when
relevant options like -pie and -shared are specified,
generate it only when the relevant relocations are actually
necessary.

Note: omitting empty __wasm_apply_data_relocs is not a problem
because the export is optional in the spec (DynamicLinking.md)
and all runtime linker implementations I'm aware of implement
it that way.  (emscripten, toywasm, wasm-tools)

Motivations:

* This possibly reduces the module size

* This is also a preparation to fix
  https://github.com/llvm/llvm-project/issues/107387,
  for which it isn't obvious if we need these relocations at the
  time of createSyntheticSymbols. (unless we introduce a new explicit
  option like --non-pie-dynamic-link.)

sbc100 commented on lld/wasm/InputChunks.cpp:
```diff
@@ -378,7 +378,7 @@ void InputChunk::generateRelocationCode(raw_ostream &os) const {
 
     Symbol *sym = file->getSymbol(rel);
-    if (!ctx.isPic && sym->isDefined())
+    if (!ctx.isPic && !sym->hasGOTIndex())
```
Is this change needed as part of this change?   I think this line could probably use a comment.   i.e. Why are we skipping these relocations?

sbc100 replied:
Should this change be differed to #108146 since it seems like a separate semantic change perhaps?

yamt replied:
this change is necessary because this function is now called regardless of unresolved policy.

sbc100 replied:
I'm finding this clauses a little hard to grok without a lot of introspection.

Maybe we could add a comment here?  Something like:

`// Runtime relocations are needed when we don't know the address of a symbol statically.`

Or better still, we could add a variable to make it more clear what is going on:

```
bool requiresRuntimeReloc = ctx.isPic || sym->hasGOTIndex();
if (!requiresRuntimeReloc)
   continue;
```

yamt replied:
> I'm finding this clauses a little hard to grok without a lot of introspection.

i agree!

> Maybe we could add a comment here? Something like:
> 
> `// Runtime relocations are needed when we don't know the address of a symbol statically.`
> 
> Or better still, we could add a variable to make it more clear what is going on:
> 
> ```
> bool requiresRuntimeReloc = ctx.isPic || sym->hasGOTIndex();
> if (!requiresRuntimeReloc)
>    continue;
> ```

applied. thank you.

sbc100 commented on lld/wasm/Writer.cpp:
```diff
@@ -1467,15 +1469,29 @@ void Writer::createApplyDataRelocationsFunction() {
     for (const OutputSegment *seg : segments)
       if (!config->sharedMemory || !seg->isTLS())
         for (const InputChunk *inSeg : seg->inputSegments)
           inSeg->generateRelocationCode(os);
```
Would it be cleaner to return a boolean from `generateRelocationCode`? 

i.e.

```
wroteRelocations |= inSeg->generateRelocationCode(os);
...
if (!wroteRelocations) {
  return;
}
```

Not a big deal but it seems more explictly/obvious.

yamt replied:
done

sbc100 commented:
How about making the title `[lld]WebAssembly] Avoid emitting empty __wasm_apply_data_relocs function`?

sbc100 commented:
I'm still not clear about the second motivation.  i.e. I'm not clear on why this is needed for the followup change.  But I'll see if I can figure it out.

yamt commented:
> How about making the title `[lld]WebAssembly] Avoid emitting empty __wasm_apply_data_relocs function`?

done

yamt committed:
clarify a bit

from sbc100
https://github.com/llvm/llvm-project/pull/109249#discussion_r1767771344

yamt commented:
can this land?
